### PR TITLE
Recyclerview support

### PR DESCRIPTION
**Pull request by zsoltk**
This is an experimental support at the moment.

Client code example in ```RecyclerViewTestActivity```.

To test in example app, set it as launcher in app-example / AndroidManifest.xml 

Relevant client code:

1. Define set of possible elements in RecyclerView
```kotlin
sealed class Item : Parcelable {
    @Parcelize object LoremIpsumItem : Item()
    @Parcelize object FooBarItem : Item()
    @Parcelize object Switcher : Item()
}
```

2. Define resolver to RoutingAction. Use ```recyclerView```, not ```attach```.
```kotlin
private val ribResolver = object : RecyclerViewRibResolver<Item> {
    override fun resolve(element: Item): RoutingAction<RibView> =
        when (element) {
            Item.LoremIpsumItem -> recyclerView { loremIpsumBuilder.build(it) }
            Item.FooBarItem -> recyclerView { fooBarBuilder.build(it) }
            Item.Switcher-> recyclerView { switcherBuilder.build(it) }
        }
}
``` 

3. Use ```RecyclerViewHostBuilder``` as any other RIB builder.

Dependencies support setting initial list of elements as well as Input stream of commands to modify later. Only ```Add``` operation is supported at the moment, others will be added later.

You can also set ```HostingStrategy``` (EAGER vs LAZY) and ```LayoutManager```.

### CODE REVIEW

File: android/app-example/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -12,7 +12,7 @@
         android:supportsRtl="true"
         android:theme="@style/AppTheme">
 
-        <activity android:name=".app.RootActivity">
+        <activity android:name=".app.RecyclerViewTestActivity">
```
Comment (CherryPerry) [inline android/app-example/src/main/AndroidManifest.xml]: > To test in example app, set it as launcher in app-example / AndroidManifest.xml

Is this change intended? Or you just forgot to switch it back to RootActivity?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostRouter.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    private val ribResolver: RecyclerViewRibResolver<T>
+): Router<Configuration, Nothing, Content, Nothing, RibView>(
+    savedInstanceState = savedInstanceState,
+    initialConfiguration = Default,
+    permanentParts = emptyList()
+) {
+    sealed class Configuration : Parcelable {
+        sealed class Content : Configuration() {
+            @Parcelize object Default : Content()
+            @Parcelize data class Item(val uuid: UUID): Content()
```
Comment (lukaville) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostRouter.kt]: maybe let's use `String` id (or custom id class) instead of `UUID` class?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        if (hostingStrategy == LAZY) {
+            router.add(configurationKey, Item(holder.uuid!!))
+        }
+        router.activate(configurationKey)
+        router.getNodes(configurationKey)!!.forEach { childNode ->
+            childNode.attachToView(holder.itemView as FrameLayout)
+        }
+    }
+
+    override fun onViewDetachedFromWindow(holder: ViewHolder) {
```
Comment (CherryPerry) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: I am not sure about this callback. Looks like it has some [issues](https://stackoverflow.com/questions/42298060/recyclerview-adapter-onviewdetachedfromwindow-never-called/51820590)? Does it work fine, when you go back from screen?

Also I checked [ViewPager2](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-master-dev:viewpager2/src/main/java/androidx/viewpager2/adapter/FragmentStateAdapter.java), that uses RecyclerView for Fragments, and they handle it via `onViewRecycled`.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import com.badoo.mvicore.element.TimeCapsule
+import com.badoo.mvicore.feature.ReducerFeature
+import com.badoo.ribs.android.recyclerview.RecyclerViewHost.Input
+import com.badoo.ribs.android.recyclerview.RecyclerViewHostFeature.State
+import com.badoo.ribs.core.routing.configuration.ConfigurationKey
+import io.reactivex.Observable
+import kotlinx.android.parcel.Parcelize
+import java.util.UUID
+
+private val timeCapsuleKey = RecyclerViewHostFeature::class.java.name
```
Comment (calimbak) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt]: Use a constant string to avoid having a random string clash with obfuscation?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            state.copy(
+                lastCommand = null
+            )
+        }
+    }
+
+    @Parcelize
+    data class State<T : Parcelable>(
+        val nextKey: Int = 1000,
+        val items: List<Entry<T>> = listOf(),
```
Comment (calimbak) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt]: emptyList()?

File: android/app-example/src/main/AndroidManifest.xml (hunk 1 above)
Comment (zsoltk) [reply android/app-example/src/main/AndroidManifest.xml]: Yes, forgot. Thanks!

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostRouter.kt (hunk 2 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostRouter.kt]: why?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt (hunk 4 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt]: Good catch, thanks.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        super.onViewDetachedFromWindow(holder)
+        val configurationKey = holder.configurationKey!! // at this point it should be bound
+        router.deactivate(configurationKey)
+        router.getNodes(configurationKey)!!.forEach { childNode ->
+            childNode.detachFromView()
+        }
+        if (hostingStrategy == LAZY) {
+            router.remove(holder.configurationKey!!)
```
[the remark this answers is not in the document]
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: Good catch, thanks.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        }
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
+        ViewHolder(
+            FrameLayout(parent.context).apply {
+                layoutParams = FrameLayout.LayoutParams(
+                    FrameLayout.LayoutParams.MATCH_PARENT,
+                    FrameLayout.LayoutParams.WRAP_CONTENT
+                )
```
[the remark this answers is not in the document]
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: Good catch, will extract to deps.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt (hunk 3 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: Good catch, will investigate further. Thanks!

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostRouter.kt (hunk 2 above)
Comment (ShikaSD) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostRouter.kt]: I wouldn't tie our implementation to UUID.
Custom `Id` class will be much more manageable if we will have to change it in the future.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            childNode.detachFromView()
+        }
+        if (hostingStrategy == LAZY) {
+            router.remove(holder.configurationKey!!)
+        }
+    }
+
+    internal fun onDestroy() {
+        items.forEach {
+            router.deactivate(it.configurationKey)
```
Comment (ShikaSD) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: You do the same a bit above (`deactivate` -> `forEach { detachFromView }` )
Maybe extract?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Router.kt [hunk 9]
```diff
@@ -97,6 +103,33 @@ abstract class Router<C : Parcelable, Permanent : C, Content : C, Overlay : C, V
...
+    }
+
+    internal fun deactivate(configurationKey: ConfigurationKey) {
+        configurationFeature.accept(
+            Deactivate(configurationKey)
+        )
+    }
+
+    internal fun getNodes(configurationKey: ConfigurationKey) =
+        (configurationFeature.state.pool[configurationKey] as? ConfigurationContext.Resolved<C>)?.nodes?.map { it.node }
```
Comment (ShikaSD) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Router.kt]: Report error if nothing found?

Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Router.kt]: Call site gets ```List<Node<*>>?``` as result, and has the responsibility to handle if it asked for something impossible.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt (hunk 8 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: Done.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt (hunk 5 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt]: Done.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+            router.add(configurationKey, Item(holder.uuid!!))
+        }
+        router.activate(configurationKey)
+        router.getNodes(configurationKey)!!.forEach { childNode ->
+            childNode.attachToView(holder.itemView as FrameLayout)
+        }
+    }
+
+    override fun onViewRecycled(holder: ViewHolder) {
+        super.onViewDetachedFromWindow(holder)
```
Comment (CherryPerry) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: Invalid `super` call.

Also if you use `onViewRecycled`, then you should also handle `onFailedToRecycleView` in the same way as `ViewPager2` does. It is valid case, because `RecyclerView` has `SimpleItemAnimator` as animator by default, and you don't set it to `null`.

File: android/app-example/src/main/java/com/badoo/ribs/example/rib/switcher/SwitcherView.kt [hunk 11]
```diff
@@ -58,7 +57,7 @@ class SwitcherViewImpl private constructor(
 
     private val menuContainer: ViewGroup = androidView.findViewById(R.id.menu_container)
     private val contentContainer: ViewGroup = androidView.findViewById(R.id.content_container)
-    private val blockerContainer: ViewGroup = androidView.findViewById(R.id.blocker_container)
+//    private val blockerContainer: ViewGroup = androidView.findViewById(R.id.blocker_container)
```
Comment (SavinMike) [inline android/app-example/src/main/java/com/badoo/ribs/example/rib/switcher/SwitcherView.kt]: Can you delete it or uncomment it

File: android/app-example/src/main/res/layout/rib_switcher.xml [hunk 12]
```diff
@@ -8,9 +8,9 @@
     android:layout_height="match_parent"
     tools:context=".app.RootActivity">
 
-    <androidx.constraintlayout.widget.ConstraintLayout
-        android:layout_width="match_parent"
-        android:layout_height="match_parent">
+    <!--<androidx.constraintlayout.widget.ConstraintLayout-->
```
Comment (SavinMike) [inline android/app-example/src/main/res/layout/rib_switcher.xml]: remove?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    override fun onViewAttachedToWindow(holder: ViewHolder) {
+        super.onViewAttachedToWindow(holder)
+        val configurationKey = holder.configurationKey!! // at this point it should be bound
+        if (hostingStrategy == LAZY) {
+            router.add(configurationKey, Item(holder.uuid!!))
+        }
+        router.activate(configurationKey)
+        router.getNodes(configurationKey)!!.forEach { childNode ->
```
Comment (SavinMike) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: why we need !! here?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        val entry = items[position]
+        holder.configurationKey = entry.configurationKey
+        holder.uuid = entry.uuid
+
+    }
+
+    override fun onViewAttachedToWindow(holder: ViewHolder) {
+        super.onViewAttachedToWindow(holder)
+        val configurationKey = holder.configurationKey!! // at this point it should be bound
```
Comment (SavinMike) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: we can avoid using so huge amount of !!, from my point of view better to remove all of them and provide more sensitive exception

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHost.kt [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        fun hostingStrategy(): HostingStrategy
+        fun initialElements(): List<T>
+        fun recyclerViewHostInput(): ObservableSource<Input<T>>
+        fun resolver(): RecyclerViewRibResolver<T>
+        fun layoutManagerFactory(): LayoutManagerFactory
+        fun viewHolderLayoutParams(): FrameLayout.LayoutParams
+
+        interface LayoutManagerFactory: (Context) -> RecyclerView.LayoutManager {
+            companion object {
+                val linear = object : LayoutManagerFactory {
```
Comment (SavinMike) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHost.kt]: I think that we shouldn't use predefined const, better to provide fun which will create factory for you, like:
```
inline fun LayoutManagerFactory(crossinline builder: (Context) -> RecyclerView.LayoutManager) =
    object : LayoutManagerFactory{
        override fun invoke(context: Context): RecyclerView.LayoutManager =
            builder(context)
    }

fun howToUse() {
    LayoutManagerFactory(::LinearLayoutManager)
}
```

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHost.kt [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+ */
+interface RecyclerViewHost : Rib {
+
+    interface Dependency<T : Parcelable> {
+        fun hostingStrategy(): HostingStrategy
+        fun initialElements(): List<T>
+        fun recyclerViewHostInput(): ObservableSource<Input<T>>
+        fun resolver(): RecyclerViewRibResolver<T>
+        fun layoutManagerFactory(): LayoutManagerFactory
+        fun viewHolderLayoutParams(): FrameLayout.LayoutParams
```
Comment (SavinMike) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHost.kt]: May be better to move it to `Customization`,  and provide default value, because almost in 99 cases you will use Match/Wrap

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+    override fun getItemCount(): Int =
+        items.size
+
+    override fun accept(state: RecyclerViewHostFeature.State<T>) {
+        items = state.items
+
+        when (state.lastCommand) {
+            null -> { /* No-op when restored from TimeCapsule or genuinely empty state */ }
+            is Input.Add -> {
+                eagerAdd(state.items.last())
```
Comment (SavinMike) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: What if we want to add N elements, or remove, or replace. Can we add more operations here?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHost.kt [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+        /**
+         * Child RIBs get created when their associated ViewHolders are attached, and get destroyed
+         * along with them
+         */
+        LAZY
+    }
+
+    sealed class Input<T : Parcelable> : Parcelable {
+        @Parcelize
+        data class Add<T : Parcelable>(val element: T): Input<T>()
```
Comment (SavinMike) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHost.kt]: What about other operations, like delete, replace and clear?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+    data class State<T : Parcelable>(
+        val nextKey: Int = 1000,
+        val items: List<Entry<T>> = emptyList(),
+        val lastCommand: Input<T>? = null
+    ) : Parcelable {
+
+        @Parcelize
+        data class Entry<T : Parcelable>(
+            val element: T,
+            val uuid: UUID = UUID.randomUUID(),
```
Comment (SavinMike) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt]: Can we use something else instead of UUID, such approach does not expandable.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+
+    class BootstrapperImpl<T : Parcelable>(
+        private val initialState: State<T>,
+        private val initialElements: List<T>
+    ) : Bootstrapper<Input<T>> {
+        override fun invoke(): Observable<Input<T>> =
+            Observable
+                .fromIterable(if (initialState.items.isEmpty()) initialElements else emptyList())
+                .map {
+                    Input.Add(it)
```
Comment (SavinMike) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt]: it's not optimal, and sometimes will provide unexpected behaviour

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+        timeCapsule.register(timeCapsuleKey) {
+            state.copy(
+                lastCommand = null
+            )
+        }
+    }
+
+    @Parcelize
+    data class State<T : Parcelable>(
+        val nextKey: Int = 1000,
```
Comment (SavinMike) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt]: why 1000?)

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostBuilder.kt [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+            input = dependency.recyclerViewHostInput(),
+            feature = feature,
+            adapter = adapter
+        )
+
+        return RecyclerViewHostNode(
+            savedInstanceState = savedInstanceState,
+            router = router,
+            interactor = interactor,
+            viewDeps =  object : RecyclerViewHostView.Dependency {
```
Comment (SavinMike) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostBuilder.kt]: Looks like adapter should be created inside of View

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostView.kt [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+
+internal class RecyclerViewHostViewImpl private constructor(
+    override val androidView: RecyclerView
+) : RecyclerViewHostView {
+
+    class Factory: RecyclerViewHostView.Factory {
+        override fun invoke(deps: Dependency): (ViewGroup) -> RecyclerViewHostView = {
+            RecyclerViewHostViewImpl(
+                androidView = RecyclerView(it.context).apply {
+                    adapter = deps.adapter()
```
Comment (SavinMike) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostView.kt]: how can we modify recyclerView? can it be part of viewDependency? Because how we can solve problem when we need to add item decorator, item Animator or change snap?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt (hunk 14 above)
Comment (CherryPerry) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: `requireNotNull(...)` with message for example.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt (hunk 13 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: because it's a map?
this method is internal and is not indented to be used in general, as the configuration pool should be accessed only by the ConfigurationFeature.
here we are sure that it's correct usage, hence the !!.

File: android/app-example/src/main/res/layout/rib_switcher.xml (hunk 12 above)
Comment (zsoltk) [reply android/app-example/src/main/res/layout/rib_switcher.xml]: I'll remove in separate ticket, as it relates to Blocker should be shown through Portal, and requires more work than this scope. Now it's quick fix to display correctly.

File: android/app-example/src/main/java/com/badoo/ribs/example/rib/switcher/SwitcherView.kt (hunk 11 above)
Comment (zsoltk) [reply android/app-example/src/main/java/com/badoo/ribs/example/rib/switcher/SwitcherView.kt]: Same as below.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt (hunk 14 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: Feel free to suggest. I'm keeping it in the meantime.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHost.kt (hunk 15 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHost.kt]: inline will not compile unfortunately, but leaving it a typealias instead of an interface will, and it will work even better in client code: 
```= ::LinearLayoutManager```

thanks!

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt (hunk 19 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt]: It's already used elsewhere and is outside the scope of this ticket. Otherwise feel free to open PR :)

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt (hunk 20 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt]: Please elaborate.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt (hunk 21 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostFeature.kt]: For debugging purposes. Otherwise has no significance.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostView.kt (hunk 23 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostView.kt]: Good catch. I can lift it out to dependency similar to LayoutManager.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostBuilder.kt (hunk 22 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostBuilder.kt]: Can't. It's used in Interactor in a binding to feature.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt (hunk 10 above)
Comment (CherryPerry) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: @zsoltk

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHost.kt [hunk 24]
```diff
@@ -19,6 +19,7 @@ interface RecyclerViewHost : Rib {
         fun initialElements(): List<T>
         fun recyclerViewHostInput(): ObservableSource<Input<T>>
         fun resolver(): RecyclerViewRibResolver<T>
+        fun recyclerViewFactory(): RecyclerViewFactory
         fun layoutManagerFactory(): LayoutManagerFactory
```
Comment (SavinMike) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHost.kt]: If we allow to create recycler, should we still pass LayoutManager?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostRouter.kt (hunk 2 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHostRouter.kt]: will do later together with other places using this, now it's not critical

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt (hunk 10 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: I missed this. Will fix!

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHost.kt (hunk 24 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/RecyclerViewHost.kt]: Right now enforces not forgetting it by accident.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt (hunk 10 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt]: Fixed in #124
